File: src/main/java/com/sleepkqq/sololeveling/mapper/DtoMapper.java
package com.sleepkqq.sololeveling.mapper;

import com.sleepkqq.sololeveling.model.UserData;
import com.sleepkqq.sololeveling.proto.user.UserInfo;
import com.sleepkqq.sololeveling.proto.user.UserRole;
import java.util.Collection;
import java.util.List;
import java.util.Locale;
import java.util.function.Function;
import one.util.streamex.StreamEx;
import org.springframework.stereotype.Component;

@Component
public class DtoMapper {

  public <T, R> List<R> mapCollection(Collection<T> collection, Function<T, R> mapper) {
    return StreamEx.of(collection).map(mapper).toList();
  }

  public UserInfo map(UserData user) {
    return UserInfo.newBuilder()
        .setId(user.getId())
        .setUsername(user.getUsername())
        .setFirstName(user.getFirstName())
        .setLastName(user.getLastName())
        .setPhotoUrl(user.getPhotoUrl())
        .setLocale(user.getLocale().toLanguageTag())
        .addAllRole(mapCollection(user.getRoles(), this::map))
        .build();
  }

  public UserData map(UserInfo userInfo) {
    return new UserData(
        userInfo.getId(),
        userInfo.getUsername(),
        userInfo.getFirstName(),
        userInfo.getLastName(),
        userInfo.getPhotoUrl(),
        Locale.forLanguageTag(userInfo.getLocale()),
        mapCollection(userInfo.getRoleList(), this::map)
    );
  }

  public UserRole map(com.sleepkqq.sololeveling.model.UserRole userRole) {
    return UserRole.valueOf(userRole.name());
  }

  public com.sleepkqq.sololeveling.model.UserRole map(UserRole userRole) {
    return com.sleepkqq.sololeveling.model.UserRole.valueOf(userRole.name());
  }
}

File: solo-leveling-ui-service/src/main/kotlin/com.sleepkqq.sololeveling/ui/mapper/DtoMapper.kt
package com.sleepkqq.sololeveling.ui.mapper

import com.sleepkqq.sololeveling.proto.user.UserInfo
import com.sleepkqq.sololeveling.proto.user.UserRole
import com.sleepkqq.sololeveling.ui.model.UserData
import org.springframework.stereotype.Component
import java.util.Locale
import java.util.UUID

@Component
class DtoMapper {

	fun map(user: UserData): UserInfo = UserInfo.newBuilder()
		.setId(user.id)
		.setUsername(user.username)
		.setFirstName(user.firstName)
		.setLastName(user.lastName)
		.setPhotoUrl(user.photoUrl)
		.setLocale(user.locale.toLanguageTag())
		.addAllRole(user.roles.map { map(it) })
		.build()

	fun map(userInfo: UserInfo): UserData = UserData(
		userInfo.id,
		userInfo.username,
		userInfo.firstName,
		userInfo.lastName,
		userInfo.photoUrl,
		Locale.forLanguageTag(userInfo.locale),
		userInfo.roleList.map { map(it) }
	)

	fun map(userRole: com.sleepkqq.sololeveling.ui.model.UserRole): UserRole {
		return UserRole.valueOf(userRole.name)
	}

	fun map(userRole: UserRole): com.sleepkqq.sololeveling.ui.model.UserRole {
		return com.sleepkqq.sololeveling.ui.model.UserRole.valueOf(userRole.name)
	}

	fun map(input: String): UUID = UUID.fromString(input)

	fun map(input: UUID): String = input.toString()
}
